# 01 – Data Cleaning and Preparation

## Project Overview

The goal of this notebook is to prepare the LinkedIn Job Postings dataset for machine learning.

Using the raw job posting data, I'll:

- Load and inspect the dataset
- Understand the available features
- Investigate missing values
- Select relevant variables for salary prediction
- Clean and preprocess the data
- Export a modeling-ready dataset for future analysis

The cleaned dataset produced in this notebook will be used throughout the remainder of the project.

In [20]:
# Import libraries 
import pandas as pd
import numpy as np
from pathlib import Path

pd.set_option("display.max_columns", None)
pd.set_option("display.max_colwidth", 100)

In [21]:
project_root = Path.cwd().parent

raw_dir = project_root / "data" / "raw"
processed_dir = project_root / "data" / "processed"

raw_dir

PosixPath('/Users/lavin/salary-prediction-ml/data/raw')

In [22]:
postings = pd.read_csv(raw_dir / "postings.csv")

In [23]:
# Initial Exploration
postings.head()

,job_id,company_name,title,description,max_salary,pay_period,location,company_id,views,med_salary,min_salary,formatted_work_type,applies,original_listed_time,remote_allowed,job_posting_url,application_url,application_type,expiry,closed_time,formatted_experience_level,skills_desc,listed_time,posting_domain,sponsored,work_type,currency,compensation_type,normalized_salary,zip_code,fips
0,921716,Corcoran Sawyer Smith,Marketing Coordinator,Job descriptionA leading real estate firm in New Jersey is seeking an administrative Marketing C...,20.0,HOURLY,"Princeton, NJ",2774458.0,20.0,NaN,17.0,Full-time,2.0,1.713398e+12,NaN,https://www.linkedin.com/jobs/view/921716/?trk=jobs_biz_prem_srch,NaN,ComplexOnsiteApply,1.715990e+12,NaN,NaN,Requirements: \n\nWe are seeking a College or Graduate Student (can also be completed with schoo...,1.713398e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,38480.0,8540.0,34021.0
1,1829192,NaN,Mental Health Therapist/Counselor,"At Aspen Therapy and Wellness , we are committed to serving clients with best practices to help ...",50.0,HOURLY,"Fort Collins, CO",NaN,1.0,NaN,30.0,Full-time,NaN,1.712858e+12,NaN,https://www.linkedin.com/jobs/view/1829192/?trk=jobs_biz_prem_srch,NaN,ComplexOnsiteApply,1.715450e+12,NaN,NaN,NaN,1.712858e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,83200.0,80521.0,8069.0
2,10998357,The National Exemplar,Assitant Restaurant Manager,The National Exemplar is accepting applications for an Assistant Restaurant Manager.\nWe offer h...,65000.0,YEARLY,"Cincinnati, OH",64896719.0,8.0,NaN,45000.0,Full-time,NaN,1.713278e+12,NaN,https://www.linkedin.com/jobs/view/10998357/?trk=jobs_biz_prem_srch,NaN,ComplexOnsiteApply,1.715870e+12,NaN,NaN,We are currently accepting resumes for FOH - Asisstant Restaurant Management with a strong focus...,1.713278e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,55000.0,45202.0,39061.0
3,23221523,"Abrams Fensterman, LLP",Senior Elder Law / Trusts and Estates Associate Attorney,Senior Associate Attorney - Elder Law / Trusts and Estates Our legal team is committed to provi...,175000.0,YEARLY,"New Hyde Park, NY",766262.0,16.0,NaN,140000.0,Full-time,NaN,1.712896e+12,NaN,https://www.linkedin.com/jobs/view/23221523/?trk=jobs_biz_prem_srch,NaN,ComplexOnsiteApply,1.715488e+12,NaN,NaN,This position requires a baseline understanding of online marketing including Search Engine Mark...,1.712896e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,157500.0,11040.0,36059.0
4,35982263,NaN,Service Technician,Looking for HVAC service tech with experience in commerical and industrial equipment. Minimum 5 ...,80000.0,YEARLY,"Burlington, IA",NaN,3.0,NaN,60000.0,Full-time,NaN,1.713452e+12,NaN,https://www.linkedin.com/jobs/view/35982263/?trk=jobs_biz_prem_srch,NaN,ComplexOnsiteApply,1.716044e+12,NaN,NaN,NaN,1.713452e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,70000.0,52601.0,19057.0


In [24]:
postings.shape

(123849, 31)

In [25]:
postings.columns.tolist()

['job_id',
 'company_name',
 'title',
 'description',
 'max_salary',
 'pay_period',
 'location',
 'company_id',
 'views',
 'med_salary',
 'min_salary',
 'formatted_work_type',
 'applies',
 'original_listed_time',
 'remote_allowed',
 'job_posting_url',
 'application_url',
 'application_type',
 'expiry',
 'closed_time',
 'formatted_experience_level',
 'skills_desc',
 'listed_time',
 'posting_domain',
 'sponsored',
 'work_type',
 'currency',
 'compensation_type',
 'normalized_salary',
 'zip_code',
 'fips']

In [26]:
# Dataset Structure
postings.info()

<class 'pandas.DataFrame'>
RangeIndex: 123849 entries, 0 to 123848
Data columns (total 31 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   job_id                      123849 non-null  int64  
 1   company_name                122130 non-null  str    
 2   title                       123849 non-null  str    
 3   description                 123842 non-null  str    
 4   max_salary                  29793 non-null   float64
 5   pay_period                  36073 non-null   str    
 6   location                    123849 non-null  str    
 7   company_id                  122132 non-null  float64
 8   views                       122160 non-null  float64
 9   med_salary                  6280 non-null    float64
 10  min_salary                  29793 non-null   float64
 11  formatted_work_type         123849 non-null  str    
 12  applies                     23320 non-null   float64
 13  original_listed_time     

In [27]:
# Missing value analysis
missing = (
    postings.isnull()
    .sum()
    .sort_values(ascending=False)
)

missing

closed_time                   122776
skills_desc                   121410
med_salary                    117569
remote_allowed                108603
applies                       100529
max_salary                     94056
min_salary                     94056
currency                       87776
compensation_type              87776
normalized_salary              87776
pay_period                     87776
posting_domain                 39968
application_url                36665
formatted_experience_level     29409
fips                           27415
zip_code                       20872
company_name                    1719
company_id                      1717
views                           1689
description                        7
work_type                          0
sponsored                          0
job_id                             0
listed_time                        0
expiry                             0
application_type                   0
original_listed_time               0
f

## Feature Selection

Not every column is useful for predicting salary.

This project focuses on variables that describe the characteristics of each job posting while excluding identifiers, URLs, and variables that would introduce data leakage.

The target variable for prediction is:

**normalized_salary**

In [28]:
# I will now create the modeling dataset

jobs = postings.copy()

jobs = jobs.dropna(subset=["normalized_salary"])

jobs.shape


model_columns = [
    "title",
    "location",
    "formatted_work_type",
    "formatted_experience_level",
    "company_name",
    "views",
    "applies",
    "normalized_salary"
]

ml_df = jobs[model_columns].copy()

ml_df.head()

,title,location,formatted_work_type,formatted_experience_level,company_name,views,applies,normalized_salary
0,Marketing Coordinator,"Princeton, NJ",Full-time,NaN,Corcoran Sawyer Smith,20.0,2.0,38480.0
1,Mental Health Therapist/Counselor,"Fort Collins, CO",Full-time,NaN,NaN,1.0,NaN,83200.0
2,Assitant Restaurant Manager,"Cincinnati, OH",Full-time,NaN,The National Exemplar,8.0,NaN,55000.0
3,Senior Elder Law / Trusts and Estates Associate Attorney,"New Hyde Park, NY",Full-time,NaN,"Abrams Fensterman, LLP",16.0,NaN,157500.0
4,Service Technician,"Burlington, IA",Full-time,NaN,NaN,3.0,NaN,70000.0


In [29]:
ml_df.isnull().sum().sort_values(ascending=False)

applies                       27249
formatted_experience_level     8119
company_name                    510
views                           488
title                             0
location                          0
formatted_work_type               0
normalized_salary                 0
dtype: int64

## Standardize Text Features

Before handling missing values, I standardize the categorical text features by removing any leading or trailing whitespace.

This helps ensure that categories differing only by extra spaces (for example, `"Data Engineer"` versus `" Data Engineer"`) are treated as the same value, reducing duplicate categories and improving the quality of the engineered features.

In [30]:
text_columns = [
    "title",
    "location",
    "formatted_work_type",
    "formatted_experience_level",
    "company_name"
]

for col in text_columns:
    ml_df[col] = ml_df[col].str.strip()

ml_df.head()

,title,location,formatted_work_type,formatted_experience_level,company_name,views,applies,normalized_salary
0,Marketing Coordinator,"Princeton, NJ",Full-time,NaN,Corcoran Sawyer Smith,20.0,2.0,38480.0
1,Mental Health Therapist/Counselor,"Fort Collins, CO",Full-time,NaN,NaN,1.0,NaN,83200.0
2,Assitant Restaurant Manager,"Cincinnati, OH",Full-time,NaN,The National Exemplar,8.0,NaN,55000.0
3,Senior Elder Law / Trusts and Estates Associate Attorney,"New Hyde Park, NY",Full-time,NaN,"Abrams Fensterman, LLP",16.0,NaN,157500.0
4,Service Technician,"Burlington, IA",Full-time,NaN,NaN,3.0,NaN,70000.0


## Data Cleaning

Prepare the selected features by handling missing values.

The preprocessing strategy includes:

- Replacing missing categorical values with `"Unknown"`
- Filling missing numerical values using the median
- Removing features with excessive missingness

In [31]:
ml_df = ml_df.drop(columns=["applies"])
ml_df["company_name"] = ml_df["company_name"].fillna("Unknown")
ml_df["formatted_experience_level"] = (
    ml_df["formatted_experience_level"]
    .fillna("Unknown")
)
ml_df["views"] = ml_df["views"].fillna(
    ml_df["views"].median()
)
ml_df.isnull().sum()

title                         0
location                      0
formatted_work_type           0
formatted_experience_level    0
company_name                  0
views                         0
normalized_salary             0
dtype: int64

## Remove Unrealistic Salary Outliers

During exploratory inspection, I found several job postings with extremely high normalized salary values, including salaries above $100 million.

These values are not realistic for the listed roles and would heavily distort exploratory analysis and model training. To reduce the impact of data-entry or normalization errors, I remove postings with normalized salaries above $1,000,000.

In [32]:
postings.loc[
    postings["title"] == "Principal Backend Engineer",
    [
        "title",
        "pay_period",
        "min_salary",
        "max_salary",
        "med_salary",
        "normalized_salary",
        "compensation_type"
    ]
]

,title,pay_period,min_salary,max_salary,med_salary,normalized_salary,compensation_type
266,Principal Backend Engineer,YEARLY,200.0,225.0,NaN,212.5,BASE_SALARY


In [33]:
postings.loc[
    postings["title"] == "Data Entry Specialist",
    [
        "title",
        "pay_period",
        "min_salary",
        "max_salary",
        "med_salary",
        "normalized_salary",
        "compensation_type"
    ]
]

,title,pay_period,min_salary,max_salary,med_salary,normalized_salary,compensation_type
723,Data Entry Specialist,HOURLY,16.0,16.0,NaN,33280.0,BASE_SALARY
1284,Data Entry Specialist,YEARLY,20.0,24.0,NaN,22.0,BASE_SALARY
12872,Data Entry Specialist,HOURLY,28.0,28.0,NaN,58240.0,BASE_SALARY
13448,Data Entry Specialist,YEARLY,48000.0,48000.0,NaN,48000.0,BASE_SALARY
23074,Data Entry Specialist,NaN,NaN,NaN,NaN,NaN,NaN
27670,Data Entry Specialist,NaN,NaN,NaN,NaN,NaN,NaN
42810,Data Entry Specialist,HOURLY,NaN,NaN,20.0,41600.0,BASE_SALARY
43971,Data Entry Specialist,HOURLY,NaN,NaN,20.0,41600.0,BASE_SALARY
45489,Data Entry Specialist,NaN,NaN,NaN,NaN,NaN,NaN
63412,Data Entry Specialist,HOURLY,NaN,NaN,18.0,37440.0,BASE_SALARY


## Remove Unrealistic Salary Values

During exploratory inspection of the salary data, I identified a small number of job postings with salary values that were clearly inconsistent with the listed positions and compensation information.

Examples included:

- Annual salaries exceeding **$100 million** for standard professional roles.
- Annual salaries below **$20,000**, including values such as **$1**, **$13**, and **$212.50** for full-time positions that were labeled as yearly compensation.

Further investigation of the original dataset showed that some of these records contained inconsistent or incorrectly normalized salary information. Because the goal of this project is to predict **annual salaries**, I removed observations with implausible salary values to improve data quality and reduce the influence of erroneous records on downstream analysis and machine learning models.

For this project, I retain annual salaries between **$20,000** and **$1,000,000**.

In [34]:
MIN_REASONABLE_SALARY = 20_000
MAX_REASONABLE_SALARY = 1_000_000

rows_before = len(ml_df)

ml_df = ml_df[
    (ml_df["normalized_salary"] >= MIN_REASONABLE_SALARY) &
    (ml_df["normalized_salary"] <= MAX_REASONABLE_SALARY)
].copy()

rows_after = len(ml_df)

print(f"Rows before filtering: {rows_before:,}")
print(f"Rows after filtering: {rows_after:,}")
print(f"Rows removed: {rows_before - rows_after:,}")

Rows before filtering: 36,073
Rows after filtering: 35,560
Rows removed: 513


In [35]:
ml_df["normalized_salary"].describe()

count     35560.000000
mean      96728.356510
std       60917.121679
min       20000.000000
25%       52340.600000
50%       82500.000000
75%      125000.000000
max      960000.000000
Name: normalized_salary, dtype: float64

In [36]:
print("Final dataset shape:", ml_df.shape)

ml_df.head()

Final dataset shape: (35560, 7)


,title,location,formatted_work_type,formatted_experience_level,company_name,views,normalized_salary
0,Marketing Coordinator,"Princeton, NJ",Full-time,Unknown,Corcoran Sawyer Smith,20.0,38480.0
1,Mental Health Therapist/Counselor,"Fort Collins, CO",Full-time,Unknown,Unknown,1.0,83200.0
2,Assitant Restaurant Manager,"Cincinnati, OH",Full-time,Unknown,The National Exemplar,8.0,55000.0
3,Senior Elder Law / Trusts and Estates Associate Attorney,"New Hyde Park, NY",Full-time,Unknown,"Abrams Fensterman, LLP",16.0,157500.0
4,Service Technician,"Burlington, IA",Full-time,Unknown,Unknown,3.0,70000.0


## Export Clean Dataset

Save the cleaned dataset for use in exploratory data analysis and machine learning modeling.

Export location:

`data/processed/clean_jobs.csv`

In [37]:
processed_dir.mkdir(parents=True, exist_ok=True)

ml_df.to_csv(
    processed_dir / "clean_jobs.csv",
    index=False
)

print("Saved cleaned dataset!")

Saved cleaned dataset!


# Key Takeaways

In this notebook, I transformed the raw LinkedIn job postings dataset into a clean dataset suitable for machine learning.

### Summary

- Started with **123,849** raw job postings.
- Selected features relevant for salary prediction.
- Removed observations without salary information.
- Handled missing categorical and numerical values.
- Investigated implausible salary values.
- Removed unrealistic salary observations caused by data inconsistencies.
- Produced a final dataset containing **35,560** job postings ready for exploratory analysis and predictive modeling.

The cleaned dataset exported in this notebook serves as the foundation for the remainder of this project.